configs/example/gem5_library/x86-global-inst-tracker.py

```diff
@@ -0,0 +1,159 @@
+    LocalInstTracker,
+)
+
+from gem5.components.boards.simple_board import SimpleBoard
+from gem5.components.cachehierarchies.classic.private_l1_cache_hierarchy import (
+    PrivateL1CacheHierarchy,
+)
+from gem5.components.memory.single_channel import SingleChannelDDR4_2400
+from gem5.components.processors.cpu_types import CPUTypes
+from gem5.components.processors.simple_processor import SimpleProcessor
+from gem5.isas import ISA
+from gem5.resources.resource import obtain_resource
+from gem5.simulate.exit_event import ExitEvent
+from gem5.simulate.simulator import Simulator
+
+cache_hierarchy = PrivateL1CacheHierarchy(
+    l1d_size="64kB",
+    l1i_size="64kB",
+)
+
+memory = SingleChannelDDR4_2400("1GB")
+
+processor = SimpleProcessor(cpu_type=CPUTypes.TIMING, num_cores=9, isa=ISA.X86)
+
+# setup instruction tracker
+
+# first, we need to create a global instruction tracker
+global_inst_tracker = GlobalInstTracker(
+    # threshold to trigger the event
+    inst_threshold=100_000_000
+)
+
+# then, we create a local instruction tracker for each core
+# and store them in a list
+all_trackers = []
+
+for core in processor.get_cores():
+    tracker = LocalInstTracker(
+        # we pass in the glboal instruction tracker to the local one
+        global_inst_tracker=global_inst_tracker,
+        # this parameter tells the tracker to start listening to instructions
+        # from the beginning of the simulation. If set to False, the tracker
+        # will
+        start_listening=True,
+    )
+    # we attach the tracker to the core
+    core.core.probeListener = tracker
+    # then store the tracker in the list
+    all_trackers.append(tracker)
+
+
+board = SimpleBoard(
+    clk_freq="1GHz",
+    processor=processor,
+    memory=memory,
+    cache_hierarchy=cache_hierarchy,
+)
+
+board.set_se_binary_workload(
+    binary=obtain_resource(resource_id="x86-matrix-multiply-omp"),
+    arguments=[100, 8],
+)
+
+
+def max_inst_handler():
+    # this handler tests the instruction tracker
+    # when it reached this function, it means that it successfully raised
+    # the ExitEvent.MAX_INSTS event
+    print("Reached MAX_INSTS with 100000000 instructions")
+    print("Changing threshold to 20000")
+    # we can change the threshold of the global instruction tracker
+    global_inst_tracker.changeThreshold(20000)
+    # we need to reset the counter of the global instruction tracker
+    # the counter does not reset automatically
+    global_inst_tracker.resetCounter()
+    m5.stats.dump()
+    m5.stats.reset()
+    yield False
+    print("Reached MAX_INSTS with 20000 instructions")
+    print("Stop listening to instructions")
+    # we can stop listening to instructions
+    for tracker in all_trackers:
+        tracker.stopListening()
+    # similarly, we can start listening to instructions again by calling:
+    #
+    # for tracker in all_trackers:
+    #     tracker.startListening()
+    m5.stats.dump()
+    m5.stats.reset()
+    yield False
```

Comment:
I understand the importance of having the exit events to be scheduled easily. I'm not against adding `scheduleInstExitAtInstsFromCurrent(<int>)` and `scheduleInstExitAtAbsoluteInst(<int>)`. But just a personal design decision, I prefer to keep the C++ side robust and let the Python side to handle all the decisions. I have a crazy idea. Since scheduling a list of exits is a common use case, is it possible to wrap it inside the standard library in [`_on_exit_events`](https://github.com/gem5/gem5/blob/stable/src/python/gem5/simulate/simulator.py#L347)? 
For example, we can have a functions `scheduleInstExitAtInstsFromCurrent(<int>)` and `scheduleInstExitAtAbsoluteInst(<int>)`  that take in the exit list from the user, and add it to the `_on_exit_events` to be called before the exit event generator/functions.

In config script:
```
scheduleInstExitAtAbsoluteInst(100000000) # First exit
scheduleInstExitAtAbsoluteInst(100000000 + 20000) # Second exit
scheduleInstExitAtAbsoluteInst(100000000 + 20000 + 20000) # Third exit
```

In `simulator.py`:
```
def rescheduleGlbalInstTracker(inst):
    self._global_inst_tracker.changeThreshold(inst)

def scheduleInstExitAtInstsFromCurrent(inst):
    if not self._global_inst_tracker:
            raise.ValueError("global inst tracker not setup")
    self.scheduleInstExitAtAbsoluteInst(inst + self._global_inst_tracker.getCurrentInst())

def scheduleInstExitAtAbsoluteInst(inst):
    if self._schedule_exits[MAX_INSTS]:
        self._schedule_exits[MAX_INSTS].add(inst)
        self._schedule_exits[MAX_INSTS].sort()
    else:
        if not self._global_inst_tracker:
            raise.ValueError("global inst tracker not setup")
        self._schedule_exits[MAX_INSTS] = {
            "param_list" = set(inst),
            "function" = rescheduleGlbalInstTracker
        }

# in the while loop in the run() function check for if need to reschedule (https://github.com/gem5/gem5/blob/stable/src/python/gem5/simulate/simulator.py#L732)

if self.__schedule_exits[exit_enum]:
    function = self._schedule_exits[exit_enum]["function"]
    param = self._schedule_exits[exit_enum]["param"].pop()
    function(param)
    if len(self._schedule_exits[exit_enum]["param"]) == 0:
        del self._schedule_exits[exit_enum]
```

The downside of this is that it's completely rely on the ExitEvent enum. It's a crazy idea. I'm open to any opinions.

Reply:
It's not a bad idea but I am against it because I don't think the Simulator should be responsible for setting up/scheduling exit events. We've done this before but I don't think it was a good idea. Eventually, somehow, to do this, you need the Simulator to inspect the simulated system and modify or "do something" to get it working. IN your example, it needs to know where to attach the probe listener, or at which cores you want to be counting, what sum over which cores you want to your global count, etc.

I think the Simulator module should be as agnostic as possible to the system it's simulating and should only be responsible for running the simulated system and handling the exit events accordingly. So, in this case I want it capable of receiving a "to instruction" exit events but I don't want it to be responsible for setting them or ordering them. 

Finally, I think logically it should be the cores that keep track of the instructions they have executed and where the user goes to state they want exit events to occur after X and Y instructions are executed. This means a core can be placed anywhere in the system, in any configuration, in any way, and it doesn't "break" the functionality of this exit events.

Reply:
I see. I changed the input type of the GlobalInstTracker from a single threshold to a list of thresholds. I also added a new setter function to add new thresholds to the tracker. This will allow us to implement the scheduling functions in the stdlib. Additionally, I updated the config script to showcase the changes.